apiserver/facades/agent/keyupdater/authorisedkeys.go

```diff
@@ -5,147 +5,226 @@ package keyupdater
-		// Consume the initial event.
-		if _, ok := <-watch.Changes(); ok {
-			results[i].NotifyWatcherId = api.resources.Register(watch)
-		} else {
-			err = watcher.EnsureErr(watch)
+
+		results[i].NotifyWatcherId, _, err = internal.EnsureRegisterWatcher[struct{}](
+			ctx, api.watcherRegistry, notifyWatcher,
+		)
+		if err != nil {
+			return params.NotifyWatchResults{}, fmt.Errorf(
+				"registering machine %q authorised keys watcher: %w",
+				machineId, err,
+			)
 		}
-		results[i].Error = apiservererrors.ServerError(err)
 	}
 	return params.NotifyWatchResults{Results: results}, nil
 }
 
 // AuthorisedKeys reports the authorised ssh keys for the specified machines.
-// The current implementation relies on global authorised keys being stored in
-// the model config and also the Juju system ssh key coming from the controllers
-// config. This will change as new user management and authorisation
-// functionality is added.
-func (api *KeyUpdaterAPI) AuthorisedKeys(ctx context.Context, arg params.Entities) (params.StringsResults, error) {
+// The current implementation fetches all keys on the model that have been
+// granted for use on a machine.
+// The following param error codes can be expected:
+// - [params.CodeTagInvalid] When a tag provided does not parse and is
+// considered invalid.
+// - [params.CodeTagKindNotSupported] When a tag has been supplied that is not a
+// machine tag.
+// - [params.CodeUnathorized] When the caller does not have permissions to get
+// the authorised keys for a requested tag.
+// - [params.CodeMachineInvalidID] When one of the machine tag's translates to a
+// invalid machine id.
+// - [params.CodeMachineNotFound] When one of the machine's does not exist.
+func (api *KeyUpdaterAPI) AuthorisedKeys(
+	ctx context.Context,
+	arg params.Entities,
+) (params.StringsResults, error) {
 	if len(arg.Entities) == 0 {
 		return params.StringsResults{}, nil
 	}
 	results := make([]params.StringsResult, len(arg.Entities))
 
-	// For now, authorised keys are global, common to all machines.
-	var keys []string
-	config, configErr := api.model.ModelConfig(ctx)
-	if configErr == nil {
-		keys = ssh.SplitAuthorisedKeys(config.AuthorizedKeys())
-	}
-
-	controllerConfig, err := api.controllerConfigService.ControllerConfig(ctx)
-	if err == nil {
-		keys = append(keys, ssh.SplitAuthorisedKeys(controllerConfig.SystemSSHKeys())...)
-	} else if err != nil {
-		configErr = stderrors.Join(configErr, err)
-	}
-
 	canRead, err := api.getCanRead()
 	if err != nil {
-		return params.StringsResults{}, err
+		return params.StringsResults{}, fmt.Errorf(
+			"checking can read status for key updater authorised keys: %w",
+			err,
+		)
 	}
+
 	for i, entity := range arg.Entities {
 		tag, err := names.ParseTag(entity.Tag)
 		if err != nil {
-			results[i].Error = apiservererrors.ServerError(err)
+			results[i].Error = apiservererrors.ParamsError(
+				params.CodeTagInvalid,
+				"cannot parse tag %q",
```

Comment:
I'd rather we didn't drop the parse error on the floor.

Reply:
Done & Done